ocrd/ocrd/processor/helpers.py

```diff
@@ -163,6 +175,142 @@ def run_cli(
+    log.debug("Processor instance %s (%s doing %s)", processor, name, otherrole)
+    t0_wall = perf_counter()
+    t0_cpu = process_time()
+
+    # Save the current working directory
+    old_cwd = os.getcwd()
+
+    # Turn input/output file groups into a comma separated string
+    input_file_grps_str = ','.join(input_file_grps)
+    output_file_grps_str = ','.join(output_file_grps)
+
+    # Set values for the processor
+    processor.input_file_grp = input_file_grps_str
+    processor.output_file_grp = output_file_grps_str
+    processor.page_id = page_id
+    processor.workspace = workspace
+
+    # Move inside the workspace (so that files in the METS can be found)
+    os.chdir(workspace.directory)
+
+    is_success = True
+    try:
+        processor.process()
+
+        # check output file groups are in METS
+        for output_file_grp in output_file_grps:
+            if output_file_grp not in workspace.mets.file_groups:
+                log.error(
+                    f'Invalid state: expected output file group "{output_file_grp}" not in METS (despite processor success)')
+    except Exception as e:
+        log.exception(e)
+        is_success = False
+    finally:
+        # Move back to the old directory
+        os.chdir(old_cwd)
+
+    t1_wall = perf_counter() - t0_wall
+    t1_cpu = process_time() - t0_cpu
+    logProfile.info(
+        "Executing processor '%s' took %fs (wall) %fs (CPU)( [--input-file-grp='%s' --output-file-grp='%s' --parameter='%s' --page-id='%s']" % (
+            ocrd_tool['executable'],
+            t1_wall,
+            t1_cpu,
+            input_file_grps_str or '',
+            output_file_grps_str or '',
+            json.dumps(processor.parameter) or '',
+            page_id or ''
+        ))
+
+    from ocrd.server.models.job import Job, StateEnum
+    job = await Job.get(job_id)
+    if is_success:
+        workspace.mets.add_agent(
+            name=name,
+            _type='OTHER',
+            othertype='SOFTWARE',
+            role='OTHER',
+            otherrole=otherrole,
+            notes=[({'option': 'input-file-grp'}, input_file_grps_str or ''),
+                   ({'option': 'output-file-grp'}, output_file_grps_str or ''),
+                   ({'option': 'parameter'}, json.dumps(processor.parameter or '')),
+                   ({'option': 'page-id'}, page_id or '')]
+        )
+        workspace.save_mets()
+
+        # Save the job status to the database
+        job.state = StateEnum.success
+    else:
+        job.state = StateEnum.failed
+    await job.save()
+
+
+@lru_cache(maxsize=32)
+def get_processor(parameter_str: str, processor_class=None):
+    """
+    Call this function to get back an instance of a processor. The results are cached based on the parameters.
+    The parameters must be passed as a string because
+    `dict <https://docs.python.org/3/tutorial/datastructures.html#dictionaries>`_ is unhashable,
+    therefore cannot be cached.
```

Comment:
Perhaps using `frozendict` like [this](https://stackoverflow.com/a/53394430/201318) might be a solution. The drawback of stringifying dict is that there is no guarantee two equivalent dicts are serialized in the same way (order of keys).